src/Sylius/Bundle/CoreBundle/EventListener/ChannelDeletionListener.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\EventListener;
+
+use Sylius\Bundle\ChannelBundle\Doctrine\ORM\ChannelRepository;
+use Sylius\Component\Channel\Model\ChannelInterface;
+use Sylius\Component\Resource\Event\ResourceEvent;
+use Sylius\Component\Resource\Exception\UnexpectedTypeException;
+
+/*
+ * Listener to prevent the deletion of last enabled channel.
+ *
+ * @author Gustavo Perdomo <gperdomor@gmail.com>
+ */
+class ChannelDeletionListener
```

Comment:
**Changing architecture?** Just a random thought - can we write a generic deletion listener? It would be great not to have to write (or copy-paste) another listener for another entity.

Moreover, there are some common checks shared with `HasEnabledEntityValidator`, it would be nice to extract it to another service and then pass only appropriate arguments here (same in `HasEnabledEntityValidator` - based on constraint).